Add terrors.IsRetryable(err error) helper

rdingwall opened the pull request:
This PR adds a new helper function `terrors.IsRetryable(err error)`, similar to `terrors.PrefixMatches()` etc.

This is intended to make it more convenient for engineers to write Kafka consumers without needing to cast `error`s to `terrors.Error`. For example:

```go
func handleKafkaEvent(
	ctx context.Context,
	event ...,
) streams.ResultWithDeadletter {

	if err := doStuff(ctx, event); err != nil {

		if terrors.IsRetryable(err) {
			// tell kafka to retry
			return streams.Failure{Error: err}
		}

		// no point retrying
		return streams.ImmediateDeadLetter{Error: err}
	}

	return streams.Success{}
}
```

Internally, `terrors.IsRetryable(err error)` calls `Error.IsRetryable()`.

Willyham commented on errors.go:
```diff
@@ -243,6 +243,15 @@ func PrefixMatches(err error, prefixParts ...string) bool {
+// IsRetryable returns true if the error is a terror and whether the error was caused by an action which can be
+// retried.
+func IsRetryable(err error) bool {
+	if terr, ok := Wrap(err, nil).(*Error); ok {
```
Instead of coercing this to a terror, we could define:

```
type interface retryable {
  Retryable() bool 
}
```

and test if the given error implements that interface. That way, we're not bound to terrors, but any other error with a `Retryable` function will also be correctly tested.

rdingwall replied:
Nice. I've implemented this in 8cb27826267eb533ee06232a8162522f5adcf28b.

Willyham reviewed (commented):
Thanks for adding this. By the way I noticed that the `ErrRateLimited` code I added isn't in the `retryableErrorCodes` and I think that maybe it should be?

rdingwall commented:
> Thanks for adding this. By the way I noticed that the `ErrRateLimited` code I added isn't in the `retryableErrorCodes` and I think that maybe it should be?

Added in 0227b14926d05571bae6c4ef64fd32ed252f3121
